### ISSUE TITLE
Issue with Looking Glass (Birdseye and IXP-Manager)

### ISSUE THREAD

**Issue by nzurku** [inex/birdseye#5]
Awesome project as aways!

Working on installing this and getting it to work with IXP-Manager.

I've come across a strange issue that I think is either related to expected BIRD output or some edge-case bug.

Our Birdseye' main page looking glass: http://rs1-ipv4.pit-ix.net/lg/protocols/bgp
Result: 
![image](https://user-images.githubusercontent.com/7988860/35238988-2c6e7d86-ff7d-11e7-8cd2-3c74d0d9800d.png)

IXP-Manager's Looking glass home page for the route server gets mangled but is half "working":
![image](https://user-images.githubusercontent.com/7988860/35239050-59e8668c-ff7d-11e7-8628-46d0350c3ebe.png)

If I click the "master" link in the last screenshot, I am presented with the rest of the looking glass working fine. It appears the list of peers/protocols is messed up in some manner.

If I run the command "/usr/bin/sudo /srv/birdseye/bin/birdc -4 -s /var/run/bird/bird.ctl show protocols" on the route servers, I am presented with the following:

![image](https://user-images.githubusercontent.com/7988860/35239131-99577c40-ff7d-11e7-8561-792ed731b81e.png)

Do I have a malformed BIRD config? the "device1" seems out of the ordinary.

**Comment by barryo** [comment]
Hi @nzurku 

Thanks for using IXP Manager and Birds Eye 😄 

Most Birds Eye issues with IXP Manager come done to a problem with selecting the appropriate configuration file for the Bird instance being called.]

Can you share the output of the following?

```
ls -la /var/run/bird
ls -la /srv/birdseye/*env
cat /srv/birdseye/*env | egrep -v '^#'
/usr/sbin/birdc -s /var/run/bird/rs1-ipv4.ctl show status
/srv/birdseye/bin/birdc -4 -s /var/run/bird/bird-rs1-ipv4.ctl show status
```

Separately, I noticed on Twitter you mentioned using ARouteserver? If you are - and while I'm sure it's a fine tool - you would ideally use IXP Manager for route server configuration generation as per:

http://docs.ixpmanager.org/features/route-servers/

The config IXP Manager creates is geared towards IXP Manager, the looking glasses and more tools we have in development. You'll also benefit from a decade of production use and experience.

**Comment by nzurku** [comment]
Thank you! Sure, here are the outputs:

> root@rs1:/home/nzurku# ls -la /var/run/bird
> total 0
> drwxrwxr-x  2 bird bird  80 Jan 18 04:00 .
> drwxr-xr-x 24 root root 820 Jan 22 15:53 ..
> srw-rw----  1 root root   0 Jan 18 04:00 bird6.ctl
> srw-rw----  1 root root   0 Jan 18 02:23 bird.ctl


> root@rs1:/home/nzurku# ls -la /srv/birdseye/*env
> -rw-r--r-- 1 www-data www-data 163 Jan 22 11:45 /srv/birdseye/birdseye-rs1-ipv4.env
> -rw-r--r-- 1 www-data www-data 165 Jan 22 11:45 /srv/birdseye/birdseye-rs1-ipv6.env

> root@rs1:/home/nzurku# cat /srv/birdseye/*env | egrep -v '^#'
> BIRDC="/usr/bin/sudo /srv/birdseye/bin/birdc -4 -s /var/run/bird/bird.ctl"
> LOOKING_GLASS_ENABLED=true
> CACHE_DRIVER=file
> LOOKING_GLASS_TITLE="PIT-IX Looking Glass"
> BIRDC="/usr/bin/sudo /srv/birdseye/bin/birdc -6 -s /var/run/bird/bird6.ctl"
> LOOKING_GLASS_ENABLED=true
> CACHE_DRIVER=file
> LOOKING_GLASS_TITLE="PIT-IX Looking Glass"


> root@rs1:/home/nzurku# /usr/sbin/birdc -s /var/run/bird/rs1-ipv4.ctl show status
> Unable to connect to server control socket (/var/run/bird/rs1-ipv4.ctl): No such file or directory


> root@rs1:/home/nzurku# /srv/birdseye/bin/birdc -4 -s /var/run/bird/bird-rs1-ipv4.ctl show status
> Bad parameters - socket does not exist

Obviously our sockets are different here which could be issue, although half of the looking glass does work which I would figure would fail.

Here are the 'show status' with the proper sockets.:

> root@rs1:/home/nzurku# /srv/birdseye/bin/birdc -4 -s /var/run/bird/bird6.ctl show status
> BIRD 1.6.3 ready.
> Access restricted
> BIRD 1.6.3
> Router ID is 206.71.141.2
> Current server time is 2018-01-22 15:59:24
> Last reboot on 2018-01-18 04:00:31
> Last reconfiguration on 2018-01-18 04:03:13
> Daemon is up and running
> root@rs1:/home/nzurku# /srv/birdseye/bin/birdc -4 -s /var/run/bird/bird.ctl show status
> BIRD 1.6.3 ready.
> Access restricted
> BIRD 1.6.3
> Router ID is 206.71.141.2
> Current server time is 2018-01-22 15:59:29
> Last reboot on 2018-01-18 02:23:35
> Last reconfiguration on 2018-01-18 04:03:13
> Daemon is up and running

**Comment by barryo** [comment]
Cool. And understood re sockets. As long as the config matches, which it seems to, you'll be fine.

Did you add the sudoer's entry for www-data?

**Comment by nzurku** [comment]
Indeed, our Saltstack install forces the following line on each RS /etc/sudoers file.

> www-data        ALL=(ALL)       NOPASSWD: /srv/birdseye/bin/birdc

The issue seems to be limited to the main page of the Birdseye looking glass and IXP-Manager LG. I would imagine it's something about our peer list causing a malformed API response?

**Comment by barryo** [comment]
Can you paste in the complete output of `show protocols`? And then for one BGP session, `show protocols all <protocol>` ?

As an aside, I notice from your output of show protocols above that they all use the master table. While I haven't seen the config, this seems wrong for a route server as each client should typically have its own RIB - see section 2.3 in [rfc7947](https://tools.ietf.org/html/rfc7947). 

This is baked into IXP Manager's template configs: https://github.com/inex/IXP-Manager/tree/master/resources/views/api/v4/router/server/bird

**Comment by nzurku** [comment]
Sure!

> root@rs1:/home/nzurku# /srv/birdseye/bin/birdc -4 -s /var/run/bird/bird.ctl show protocols
> BIRD 1.6.3 ready.
> Access restricted
> name     proto    table    state  since       info
> device1  Device   master   up     2018-01-18
> AS13335_1 BGP      master   up     01:17:55    Established
> AS17054_1 BGP      master   start  2018-01-18  Passive
> AS20326_1 BGP      master   up     2018-01-18  Established

> root@rs1:/home/nzurku# /srv/birdseye/bin/birdc -4 -s /var/run/bird/bird.ctl show protocols all AS20326_1
> BIRD 1.6.3 ready.
> Access restricted
> name     proto    table    state  since       info
> AS20326_1 BGP      master   up     2018-01-18  Established
>   Description:    TeraSwitch, Inc.
>   Preference:     100
>   Input filter:   receive_from_AS20326_1
>   Output filter:  announce_to_AS20326_1
>   Import limit:   3
>     Action:       disable
>   Routes:         3 imported, 0 filtered, 1 exported, 3 preferred
>   Route change stats:     received   rejected   filtered    ignored   accepted
>     Import updates:              3          0          0          0          3
>     Import withdraws:            0          0        ---          0          0
>     Export updates:              5          3          0        ---          2
>     Export withdraws:            1        ---        ---        ---          1
>   BGP state:          Established
>     Neighbor address: 206.71.141.10
>     Neighbor AS:      20326
>     Neighbor ID:      74.118.136.29
>     Neighbor caps:    refresh restart-aware AS4
>     Session:          external route-server AS4
>     Source address:   206.71.141.2
>     Route limit:      3/3
>     Hold timer:       60/90
>     Keepalive timer:  25/30

I will look into the master table issue - is that possibly related to this?

Thanks!

**Comment by barryo** [comment]
Hi @nzurku 

> Can you paste in the complete output of show protocols? And then for one BGP session, show protocols all <protocol> ?

Nothing jumps out at me as an issue here.

Before I look deeper, are you still having issues?

**Comment by nzurku** [comment]
Indeed, we still have this issue. My guess is the fact that we use that single master table is somehow related.

If this is an edge-case and time consuming, we could try another looking glass project. I know our use of Arouteserver likely falls out of usual operation of IXP-Manager projects.
